ldapRealm broke using JDK11.0.12+7

## Environment Details
* GlassFish Version (6.2.5):
* JDK version:11.0.12+7
* OS: Linux 
* Database:
----------
## Problem Description
When configuring the ldapRealm in GF6 using JDK 11.0.12 , it will wont work. 
The same configuration work fine with GF4.1.2 and JDK8 

## Steps to reproduce
Configure an application to use ldapRealm using GF6.5.2 and JDK 11.0.12 and try to log in. 

## stack trace: 
[glassfish 6.2] [SEVERE] [jakarta.enterprise.system.core.security] [tid: _ThreadID=47 _ThreadName=http-listener-2(4)] [timeMillis: 1661340894726] [levelValue: 1000] [[
  ldaprealm.exception
javax.security.auth.login.LoginException: javax.naming.CommunicationException: ldap-server-name:636 [Root exception is java.net.SocketException: Unconnected sockets not implemented]
        at com.sun.enterprise.security.auth.realm.ldap.LDAPRealm.findAndBind(LDAPRealm.java:483)
        at com.sun.enterprise.security.auth.login.LDAPLoginModule.authenticate(LDAPLoginModule.java:74)
        at com.sun.enterprise.security.auth.login.PasswordLoginModule.authenticateUser(PasswordLoginModule.java:80)
        at com.sun.enterprise.security.BasePasswordLoginModule.login(BasePasswordLoginModule.java:114)
        at java.base/javax.security.auth.login.LoginContext.invoke(LoginContext.java:747)
        at java.base/javax.security.auth.login.LoginContext$4.run(LoginContext.java:672)
        at java.base/javax.security.auth.login.LoginContext$4.run(LoginContext.java:670)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.base/javax.security.auth.login.LoginContext.invokePriv(LoginContext.java:670)
        at java.base/javax.security.auth.login.LoginContext.login(LoginContext.java:581)
        at com.sun.enterprise.security.auth.login.LoginContextDriver.doPasswordLogin(LoginContextDriver.java:327)
        at com.sun.enterprise.security.auth.login.LoginContextDriver.login(LoginContextDriver.java:199)
        at com.sun.enterprise.security.auth.login.LoginContextDriver.login(LoginContextDriver.java:124)
        at com.sun.web.security.RealmAdapter.authenticate(RealmAdapter.java:852)
        at com.sun.web.security.RealmAdapter.authenticate(RealmAdapter.java:552)
        at org.apache.catalina.authenticator.FormAuthenticator.authenticate(FormAuthenticator.java:237)
        at org.apache.catalina.authenticator.AuthenticatorBase.processSecurityCheck(AuthenticatorBase.java:963)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:549)
        at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:579)
        at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:550)
        at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:75)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:114)
        at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:332)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:199)
        at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:440)
        at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:144)
        at org.glassfish.grizzly.http.server.HttpHandler$1.run(HttpHandler.java:190)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:535)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:515)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: javax.naming.CommunicationException:  ldap-server-name:636 [Root exception is java.net.SocketException: Unconnected sockets not implemented]
        at java.naming/com.sun.jndi.ldap.Connection.<init>(Connection.java:252)
        at java.naming/com.sun.jndi.ldap.LdapClient.<init>(LdapClient.java:137)
        at java.naming/com.sun.jndi.ldap.LdapClientFactory.createPooledConnection(LdapClientFactory.java:71)
        at java.naming/com.sun.jndi.ldap.pool.Connections.createConnection(Connections.java:185)
        at java.naming/com.sun.jndi.ldap.pool.Connections.getAvailableConnection(Connections.java:151)
        at java.naming/com.sun.jndi.ldap.pool.Pool.getOrCreatePooledConnection(Pool.java:189)
        at java.naming/com.sun.jndi.ldap.pool.Pool.getPooledConnection(Pool.java:148)
        at java.naming/com.sun.jndi.ldap.LdapPoolManager.getLdapClient(LdapPoolManager.java:340)
        at java.naming/com.sun.jndi.ldap.LdapClient.getInstance(LdapClient.java:1608)
        at java.naming/com.sun.jndi.ldap.LdapCtx.connect(LdapCtx.java:2847)
        at java.naming/com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:348)
        at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxFromUrl(LdapCtxFactory.java:262)
        at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:226)
        at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:280)
        at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:185)
        at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:115)
        at java.naming/javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:730)
        at java.naming/javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:305)
        at java.naming/javax.naming.InitialContext.init(InitialContext.java:236)
        at java.naming/javax.naming.InitialContext.<init>(InitialContext.java:208)
        at java.naming/javax.naming.directory.InitialDirContext.<init>(InitialDirContext.java:101)
        at com.sun.enterprise.security.auth.realm.ldap.LDAPRealm.findAndBind(LDAPRealm.java:451)
        ... 29 more
Caused by: java.net.SocketException: Unconnected sockets not implemented
        at java.base/javax.net.SocketFactory.createSocket(SocketFactory.java:125)
        at java.naming/com.sun.jndi.ldap.Connection.createSocket(Connection.java:303)
        at java.naming/com.sun.jndi.ldap.Connection.<init>(Connection.java:231)
        ... 50 more
Caused by: java.lang.UnsupportedOperationException
        at java.base/javax.net.SocketFactory.createSocket(SocketFactory.java:123)
        ... 52 more
]]

## Impact of Issue
## info: 
This issue seem to have with certen JDK versions. I worked fine when i tried JDK 11.0.14+9, but with JDK11.0.16 it is still exist.

--- a/appserver/security/webintegration/src/main/java/com/sun/web/security/SSLSocketFactory.java
+++ b/appserver/security/webintegration/src/main/java/com/sun/web/security/SSLSocketFactory.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
  * Copyright (c) 1997, 2021 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
@@ -31,16 +32,12 @@ import javax.net.ssl.SSLServerSocket;
 import javax.net.ssl.TrustManager;
 import javax.net.ssl.X509KeyManager;
 
-import org.glassfish.hk2.api.ServiceLocator;
+import org.apache.catalina.net.ServerSocketFactory;
 import org.glassfish.internal.api.Globals;
 import org.glassfish.security.common.SharedSecureRandomImpl;
 
-//V3:Commented import com.sun.enterprise.ServerConfiguration;
-//V3:Commented import com.sun.web.server.*;
-//V3:Commented import com.sun.enterprise.server.J2EEServer;
 import com.sun.enterprise.security.ssl.J2EEKeyManager;
 import com.sun.enterprise.security.ssl.SSLUtils;
-import com.sun.enterprise.util.LocalStringManagerImpl;
 import com.sun.logging.LogDomains;
 
 /**
@@ -50,14 +47,11 @@ import com.sun.logging.LogDomains;
  * @author Vivek Nagar
  * @author Harpreet Singh
  */
-// TODO: this should become a HK2 component
-public class SSLSocketFactory implements org.apache.catalina.net.ServerSocketFactory {
+public class SSLSocketFactory implements ServerSocketFactory {
 
     static Logger _logger = LogDomains.getLogger(SSLSocketFactory.class, LogDomains.WEB_LOGGER);
 
     private static final boolean clientAuth = false;
-
-    private static LocalStringManagerImpl localStrings = new LocalStringManagerImpl(SSLSocketFactory.class);
 
     private SSLContext context;
     private javax.net.ssl.SSLServerSocketFactory factory;
@@ -68,7 +62,7 @@ public class SSLSocketFactory implements org.apache.catalina.net.ServerSocketFac
 
     // XXX initStoresAtStartup may call more than once, should clean up later
     // copied from SSLUtils : V3 to break dependency of this SSLUtils on this Class.
-    private static boolean initialized = false;
+    private static boolean initialized;
 
     /**
      * Create the SSL socket factory. Initialize the key managers and trust managers which are passed to the SSL context.
@@ -104,18 +98,8 @@ public class SSLSocketFactory implements org.apache.catalina.net.ServerSocketFac
     public ServerSocket createSocket(int port) throws IOException {
         SSLServerSocket socket = (SSLServerSocket) factory.createServerSocket(port);
         init(socket);
-        return socket;
-    }
 
-    /**
-     * Specify whether the server will require client authentication.
-     *
-     * @param socket the SSL server socket.
-     */
-    private void init(SSLServerSocket socket) {
-        // Some initialization goes here.....
-        // socket.setEnabledCipherSuites(cipherSuites);
-        socket.setNeedClientAuth(clientAuth);
+        return socket;
     }
 
     /**
@@ -128,6 +112,7 @@ public class SSLSocketFactory implements org.apache.catalina.net.ServerSocketFac
     public ServerSocket createSocket(int port, int backlog) throws IOException {
         SSLServerSocket socket = (SSLServerSocket) factory.createServerSocket(port, backlog);
         init(socket);
+
         return socket;
     }
 
@@ -141,6 +126,7 @@ public class SSLSocketFactory implements org.apache.catalina.net.ServerSocketFac
     public ServerSocket createSocket(int port, int backlog, InetAddress ifAddress) throws IOException {
         SSLServerSocket socket = (SSLServerSocket) factory.createServerSocket(port, backlog, ifAddress);
         init(socket);
+
         return socket;
     }
 
@@ -155,8 +141,7 @@ public class SSLSocketFactory implements org.apache.catalina.net.ServerSocketFac
             return;
         }
 
-        ServiceLocator habitat = Globals.getDefaultHabitat();
-        SSLUtils sslUtils = habitat.getService(SSLUtils.class);
+        SSLUtils sslUtils = Globals.getDefaultHabitat().getService(SSLUtils.class);
 
         keyManagers = sslUtils.getKeyManagers();
         trustManagers = sslUtils.getTrustManagers();
@@ -171,8 +156,18 @@ public class SSLSocketFactory implements org.apache.catalina.net.ServerSocketFac
                 keyManagers[i] = new J2EEKeyManager((X509KeyManager) keyManagers[i], keyAlias);
             }
         }
+
         sslContext.init(keyManagers, sslUtils.getTrustManagers(), null);
         HttpsURLConnection.setDefaultSSLSocketFactory(sslContext.getSocketFactory());
         initialized = true;
+    }
+
+    /**
+     * Specify whether the server will require client authentication.
+     *
+     * @param socket the SSL server socket.
+     */
+    private void init(SSLServerSocket socket) {
+        socket.setNeedClientAuth(clientAuth);
     }
 }

--- a/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/login/FileLoginModule.java
+++ b/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/login/FileLoginModule.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
  * Copyright (c) 1997, 2021 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
@@ -16,7 +17,7 @@
 
 package com.sun.enterprise.security.auth.login;
 
-import java.util.logging.Level;
+import static java.util.logging.Level.FINE;
 
 import javax.security.auth.login.LoginException;
 
@@ -48,17 +49,16 @@ public class FileLoginModule extends PasswordLoginModule {
         }
         FileRealm fileRealm = (FileRealm) _currentRealm;
 
-        String[] grpList = fileRealm.authenticate(_username, getPasswordChar());
+        String[] groups = fileRealm.authenticate(_username, getPasswordChar());
 
-        if (grpList == null) { // JAAS behavior
-            String msg = sm.getString("filelm.faillogin", _username);
-            throw new LoginException(msg);
+        if (groups == null) { // JAAS behavior
+            throw new LoginException(sm.getString("filelm.faillogin", _username));
         }
 
-        if (_logger.isLoggable(Level.FINE)) {
-            _logger.log(Level.FINE, "File login succeeded for: " + _username);
+        if (_logger.isLoggable(FINE)) {
+            _logger.log(FINE, "File login succeeded for: " + _username);
         }
 
-        commitAuthentication(_username, getPasswordChar(), _currentRealm, grpList);
+        commitAuthentication(_username, getPasswordChar(), _currentRealm, groups);
     }
 }

--- a/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/login/LDAPLoginModule.java
+++ b/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/login/LDAPLoginModule.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
  * Copyright (c) 1997, 2021 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
@@ -15,6 +16,8 @@
  */
 
 package com.sun.enterprise.security.auth.login;
+
+import static com.sun.enterprise.util.Utility.isEmpty;
 
 import javax.security.auth.login.LoginException;
 
@@ -61,21 +64,20 @@ public class LDAPLoginModule extends PasswordLoginModule {
         }
         _ldapRealm = (LDAPRealm) _currentRealm;
 
-        // enforce that password cannot be empty.
+        // Enforce that password cannot be empty.
         // ldap may grant login on empty password!
-        if (getPasswordChar() == null || getPasswordChar().length == 0) {
-            String msg = sm.getString("ldaplm.emptypassword", _username);
-            throw new LoginException(msg);
+        if (isEmpty(getPasswordChar())) {
+            throw new LoginException(sm.getString("ldaplm.emptypassword", _username));
         }
 
         String mode = _currentRealm.getProperty(LDAPRealm.PARAM_MODE);
 
-        if (LDAPRealm.MODE_FIND_BIND.equals(mode)) {
-            String[] grpList = _ldapRealm.findAndBind(_username, getPasswordChar());
-            commitAuthentication(_username, getPasswordChar(), _currentRealm, grpList);
-        } else {
-            String msg = sm.getString("ldaplm.badmode", mode);
-            throw new LoginException(msg);
+        if (!LDAPRealm.MODE_FIND_BIND.equals(mode)) {
+            throw new LoginException(sm.getString("ldaplm.badmode", mode));
         }
+
+        String[] groups = _ldapRealm.findAndBind(_username, getPasswordChar());
+
+        commitAuthentication(_username, getPasswordChar(), _currentRealm, groups);
     }
 }

--- a/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/login/PasswordLoginModule.java
+++ b/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/login/PasswordLoginModule.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
  * Copyright (c) 1997, 2021 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
@@ -26,7 +27,6 @@ import com.sun.enterprise.security.auth.realm.Realm;
  * deprecation.
  *
  */
-@Deprecated
 public abstract class PasswordLoginModule extends BasePasswordLoginModule {
 
     /**

--- a/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/realm/ldap/CustomSocketFactory.java
+++ b/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/realm/ldap/CustomSocketFactory.java
@@ -70,6 +70,11 @@ public class CustomSocketFactory extends SocketFactory implements Comparator<Soc
     }
 
     @Override
+    public Socket createSocket() throws IOException {
+        return socketFactory.createSocket();
+    }
+
+    @Override
     public Socket createSocket(String host, int port) throws IOException, UnknownHostException {
         return socketFactory.createSocket(host, port);
     }

--- a/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/realm/ldap/CustomSocketFactory.java
+++ b/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/realm/ldap/CustomSocketFactory.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
  * Copyright (c) 1997, 2021 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
@@ -16,12 +17,14 @@
 
 package com.sun.enterprise.security.auth.realm.ldap;
 
+import static com.sun.enterprise.security.SecurityLoggerInfo.securityExceptionError;
+import static java.util.logging.Level.WARNING;
+
 import java.io.IOException;
 import java.net.InetAddress;
 import java.net.Socket;
 import java.net.UnknownHostException;
 import java.util.Comparator;
-import java.util.logging.Level;
 import java.util.logging.Logger;
 
 import javax.net.SocketFactory;
@@ -44,55 +47,46 @@ import com.sun.enterprise.util.i18n.StringManager;
  *
  */
 public class CustomSocketFactory extends SocketFactory implements Comparator<SocketFactory> {
-    private SocketFactory socketFactory;
 
     public static final String SSL = "SSL";
+
     protected static final Logger _logger = SecurityLoggerInfo.getLogger();
     protected static final StringManager sm = StringManager.getManager(CustomSocketFactory.class);
+
     private static final CustomSocketFactory customSocketFactory = new CustomSocketFactory();
 
+    private SocketFactory socketFactory;
+
     public CustomSocketFactory() {
-        SSLUtils sslUtils = Globals.getDefaultHabitat().getService(SSLUtils.class);
-        SSLContext sc = null;
+        SSLUtils sslUtils = Globals.get(SSLUtils.class);
+
         try {
-            sc = SSLContext.getInstance(SSL);
-            sc.init(sslUtils.getKeyManagers(), sslUtils.getTrustManagers(), SharedSecureRandom.get());
-            socketFactory = sc.getSocketFactory();
+            SSLContext sslContext = SSLContext.getInstance(SSL);
+            sslContext.init(sslUtils.getKeyManagers(), sslUtils.getTrustManagers(), SharedSecureRandom.get());
+            socketFactory = sslContext.getSocketFactory();
         } catch (Exception ex) {
-            _logger.log(Level.WARNING, SecurityLoggerInfo.securityExceptionError, ex);
+            _logger.log(WARNING, securityExceptionError, ex);
         }
     }
 
-    /**
-     * @see javax.net.SocketFactory#createSocket(java.lang.String, int)
-     */
     @Override
-    public Socket createSocket(String arg0, int arg1) throws IOException, UnknownHostException {
-        return socketFactory.createSocket(arg0, arg1);
+    public Socket createSocket(String host, int port) throws IOException, UnknownHostException {
+        return socketFactory.createSocket(host, port);
     }
 
-    /**
-     * @see javax.net.SocketFactory#createSocket(java.net.InetAddress, int)
-     */
     @Override
-    public Socket createSocket(InetAddress arg0, int arg1) throws IOException {
-        return socketFactory.createSocket(arg0, arg1);
+    public Socket createSocket(InetAddress host, int port) throws IOException {
+        return socketFactory.createSocket(host, port);
     }
 
-    /**
-     * @see javax.net.SocketFactory#createSocket(java.lang.String, int, java.net.InetAddress, int)
-     */
     @Override
-    public Socket createSocket(String arg0, int arg1, InetAddress arg2, int arg3) throws IOException, UnknownHostException {
-        return socketFactory.createSocket(arg0, arg1, arg2, arg3);
+    public Socket createSocket(String host, int port, InetAddress localHost, int localPort) throws IOException, UnknownHostException {
+        return socketFactory.createSocket(host, port, localHost, localPort);
     }
 
-    /**
-     * @see javax.net.SocketFactory#createSocket(java.net.InetAddress, int, java.net.InetAddress, int)
-     */
     @Override
-    public Socket createSocket(InetAddress arg0, int arg1, InetAddress arg2, int arg3) throws IOException {
-        return socketFactory.createSocket(arg0, arg1, arg2, arg3);
+    public Socket createSocket(InetAddress address, int port, InetAddress localAddress, int localPort) throws IOException {
+        return socketFactory.createSocket(address, port, localAddress, localPort);
     }
 
     @Override

--- a/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/realm/ldap/LDAPRealm.java
+++ b/nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/realm/ldap/LDAPRealm.java
@@ -17,12 +17,13 @@
 
 package com.sun.enterprise.security.auth.realm.ldap;
 
-import com.sun.enterprise.security.BaseRealm;
-import com.sun.enterprise.security.auth.realm.BadRealmException;
-import com.sun.enterprise.security.auth.realm.IASRealm;
-import com.sun.enterprise.security.auth.realm.InvalidOperationException;
-import com.sun.enterprise.security.auth.realm.NoSuchRealmException;
-import com.sun.enterprise.security.auth.realm.NoSuchUserException;
+import static com.sun.enterprise.util.Utility.isAnyNull;
+import static java.util.logging.Level.FINE;
+import static java.util.logging.Level.FINEST;
+import static java.util.logging.Level.WARNING;
+import static javax.naming.Context.SECURITY_CREDENTIALS;
+import static javax.naming.Context.SECURITY_PRINCIPAL;
+import static javax.naming.directory.SearchControls.SUBTREE_SCOPE;
 
 import java.util.ArrayList;
 import java.util.Collections;
@@ -51,6 +52,13 @@ import javax.security.auth.login.LoginException;
 
 import org.glassfish.internal.api.RelativePathResolver;
 import org.jvnet.hk2.annotations.Service;
+
+import com.sun.enterprise.security.BaseRealm;
+import com.sun.enterprise.security.auth.realm.BadRealmException;
+import com.sun.enterprise.security.auth.realm.IASRealm;
+import com.sun.enterprise.security.auth.realm.InvalidOperationException;
+import com.sun.enterprise.security.auth.realm.NoSuchRealmException;
+import com.sun.enterprise.security.auth.realm.NoSuchUserException;
 
 /**
  * Realm wrapper for supporting LDAP authentication.
@@ -134,7 +142,8 @@ public final class LDAPRealm extends IASRealm {
     private static final String SUN_JNDI_POOL_ = "com.sun.jndi.ldap.connect.pool.";
     private static final String SUN_JNDI_POOL_PROTOCOL = "com.sun.jndi.ldap.connect.pool.protocol";
     private static final String SUN_JNDI_POOL_MAXSIZE = "com.sun.jndi.ldap.connect.pool.maxsize";
-    // dynamic group related properties.
+
+    // Dynamic group related properties.
     private static final String DYNAMIC_GROUP_OBJECT_FACTORY = "com.sun.jndi.ldap.obj.LdapGroupFactory";
     public static final String DYNAMIC_GROUP_FACTORY_OBJECT_PROPERTY = "java.naming.factory.object";
     private static final String DYNAMIC_GROUP_STATE_FACTORY = "com.sun.jndi.ldap.obj.LdapGroupFactory";
@@ -164,14 +173,15 @@ public final class LDAPRealm extends IASRealm {
     @Override
     public synchronized void init(Properties props) throws BadRealmException, NoSuchRealmException {
         super.init(props);
+
         String url = props.getProperty(PARAM_DIRURL);
         String dn = props.getProperty(PARAM_USERDN);
         String jaasCtx = props.getProperty(BaseRealm.JAAS_CONTEXT_PARAM);
 
-        if (url == null || dn == null || jaasCtx == null) {
-            String msg = sm.getString("ldaprealm.badconfig", url, dn, jaasCtx);
-            throw new BadRealmException(msg);
+        if (isAnyNull(url, dn, jaasCtx)) {
+            throw new BadRealmException(sm.getString("ldaprealm.badconfig", url, dn, jaasCtx));
         }
+
         this.setProperty(PARAM_DIRURL, url);
         ldapBindProps.setProperty(Context.PROVIDER_URL, url);
         this.setProperty(PARAM_USERDN, dn);
@@ -179,9 +189,9 @@ public final class LDAPRealm extends IASRealm {
 
         String mode = props.getProperty(PARAM_MODE, MODE_DEFAULT);
         if (!MODE_DEFAULT.equals(mode)) {
-            String msg = sm.getString("ldaprealm.badmode", mode);
-            throw new BadRealmException(msg);
+            throw new BadRealmException(sm.getString("ldaprealm.badmode", mode));
         }
+
         this.setProperty(PARAM_MODE, mode);
 
         String ctxF = props.getProperty(PARAM_JNDICF, JNDICF_DEFAULT);
@@ -217,24 +227,25 @@ public final class LDAPRealm extends IASRealm {
         String bindDN = props.getProperty(PARAM_BINDDN);
         if (bindDN != null) {
             this.setProperty(PARAM_BINDDN, bindDN);
-            ldapBindProps.setProperty(Context.SECURITY_PRINCIPAL, bindDN);
+            ldapBindProps.setProperty(SECURITY_PRINCIPAL, bindDN);
         }
+
         String bindPWD = props.getProperty(PARAM_BINDPWD);
         if (bindPWD != null) {
-            //If the passwors is aliased, de-alias it
+            //If the password is aliased, de-alias it
             try {
                 bindPWD = RelativePathResolver.getRealPasswordFromAlias(bindPWD);
             } catch (Exception ex) {
-                _logger.log(Level.WARNING, "ldaprealm.pwd.dealiasing.failed", ex);
+                _logger.log(WARNING, "ldaprealm.pwd.dealiasing.failed", ex);
             }
+
             this.setProperty(PARAM_BINDPWD, bindPWD);
-            ldapBindProps.setProperty(Context.SECURITY_CREDENTIALS, bindPWD);
+            ldapBindProps.setProperty(SECURITY_CREDENTIALS, bindPWD);
         }
 
         Set<String> penum = props.stringPropertyNames();
         for (String propName : penum) {
-            if (propName.startsWith("java.naming.") || propName.startsWith("javax.security.")
-                || propName.startsWith("com.sun.jndi.ldap.")) {
+            if (propName.startsWith("java.naming.") || propName.startsWith("javax.security.") || propName.startsWith("com.sun.jndi.ldap.")) {
                 ldapBindProps.setProperty(propName, props.getProperty(propName));
             } else if (propName.startsWith(SUN_JNDI_POOL_) && !SUN_JNDI_POOL_MAXSIZE.equals(propName)) {
                 if (System.getProperty(propName) == null) {
@@ -245,7 +256,8 @@ public final class LDAPRealm extends IASRealm {
 
         String poolSize = Integer.getInteger(PARAM_POOLSIZE, POOLSIZE_DEFAULT).toString();
         String sunPoolSizeStr = props.getProperty(SUN_JNDI_POOL_MAXSIZE, poolSize);
-        //Precedence rule: SUN_JNDI_POOL_MAXSIZE > PARAM_POOLSIZE > POOLSIZE_DEFAULT
+
+        // Precedence rule: SUN_JNDI_POOL_MAXSIZE > PARAM_POOLSIZE > POOLSIZE_DEFAULT
         try {
             sunPoolSizeStr = Integer.valueOf(sunPoolSizeStr).toString();
         } catch (Exception ex) {
@@ -264,15 +276,15 @@ public final class LDAPRealm extends IASRealm {
             if (System.getProperty(SUN_JNDI_POOL_PROTOCOL) == null) {
                 System.setProperty(SUN_JNDI_POOL_PROTOCOL, DEFAULT_POOL_PROTOCOL);
             }
-            if (_logger.isLoggable(Level.FINE)) {
-                _logger.log(Level.FINE, "LDAPRealm : Using custom socket factory for SSL with pooling");
+            if (_logger.isLoggable(FINE)) {
+                _logger.log(FINE, "LDAPRealm : Using custom socket factory for SSL with pooling");
             }
         }
 
-        if (_logger.isLoggable(Level.FINE)) {
+        if (_logger.isLoggable(FINE)) {
             Properties tempProps = (Properties) ldapBindProps.clone();
-            tempProps.remove(Context.SECURITY_CREDENTIALS);
-            _logger.log(Level.FINE, "LDAPRealm : " + tempProps);
+            tempProps.remove(SECURITY_CREDENTIALS);
+            _logger.log(FINE, "LDAPRealm : " + tempProps);
         }
 
         groupCache = new HashMap<>();
@@ -290,20 +302,21 @@ public final class LDAPRealm extends IASRealm {
         return AUTH_TYPE;
     }
 
-    private String[] addMappedGroupNames(String[] grpList) {
-
+    private String[] addMappedGroupNames(String[] groups) {
         if (groupMapper == null) {
-            return grpList;
+            return groups;
         }
+
         ArrayList<String> finalresult = new ArrayList<>();
-        for (String grp : grpList) {
+        for (String group : groups) {
             ArrayList<String> result = new ArrayList<>();
-            groupMapper.getMappedGroups(grp, result);
-            finalresult.add(grp);
+            groupMapper.getMappedGroups(group, result);
+            finalresult.add(group);
             if (!result.isEmpty()) {
                 finalresult.addAll(result);
             }
         }
+
         return finalresult.toArray(new String[finalresult.size()]);
     }
 
@@ -316,7 +329,7 @@ public final class LDAPRealm extends IASRealm {
     }
 
     private List<String> getGroups(String userDN) {
-        //no authentication has happened through the realm.
+        // No authentication has happened through the realm.
         DirContext ctx = null;
         String srcFilter = null;
 
@@ -336,8 +349,8 @@ public final class LDAPRealm extends IASRealm {
             }
 
             if (_username == null && userDN != null && userDN.startsWith("uid")) {
-                //handle uid=XXX here where cn is not present
-                //TODO :maybe there is a better way to handle this??
+                // Handle uid=XXX here where cn is not present
+                // TODO :maybe there is a better way to handle this??
                 int first = userDN.indexOf("uid=");
                 int last = userDN.indexOf(",");
                 if (first != -1 && last != -1) {
@@ -345,32 +358,35 @@ public final class LDAPRealm extends IASRealm {
                 }
 
             }
-            StringBuffer sb = new StringBuffer(getProperty(PARAM_GRP_SEARCH_FILTER));
-            StringBuffer dynSb = new StringBuffer(getProperty(PARAM_DYNAMIC_GRP_FILTER));
-            substitute(sb, SUBST_SUBJECT_NAME, _username);
-            substitute(sb, SUBST_SUBJECT_DN, userDN);
-            substitute(dynSb, SUBST_SUBJECT_NAME, _username);
-            substitute(dynSb, SUBST_SUBJECT_DN, userDN);
+            StringBuilder searchFilterBuilder = new StringBuilder(getProperty(PARAM_GRP_SEARCH_FILTER));
+            StringBuilder dynamicFilterBuilder = new StringBuilder(getProperty(PARAM_DYNAMIC_GRP_FILTER));
 
-            srcFilter = sb.toString();
-            dynFilter = dynSb.toString();
+            substitute(searchFilterBuilder, SUBST_SUBJECT_NAME, _username);
+            substitute(searchFilterBuilder, SUBST_SUBJECT_DN, userDN);
+            substitute(dynamicFilterBuilder, SUBST_SUBJECT_NAME, _username);
+            substitute(dynamicFilterBuilder, SUBST_SUBJECT_DN, userDN);
+
+            srcFilter = searchFilterBuilder.toString();
+            dynFilter = dynamicFilterBuilder.toString();
             List<String> groupsList = new ArrayList<>();
             groupsList.addAll(groupSearch(ctx, getProperty(PARAM_GRPDN), srcFilter, getProperty(PARAM_GRP_TARGET)));
-            // search filter is constructed internally as
-            // as a groupofURLS
+
+            // Search filter is constructed internally as as a groupofURLS
             groupsList.addAll(dynamicGroupSearch(ctx, getProperty(PARAM_GRPDN), dynMember, dynFilter, getProperty(PARAM_GRP_TARGET)));
+
             return groupsList;
         } catch (Exception e) {
-            _logger.log(Level.WARNING, "ldaprealm.groupsearcherror", e);
+            _logger.log(WARNING, "ldaprealm.groupsearcherror", e);
         } finally {
             if (ctx != null) {
                 try {
                     ctx.close();
                 } catch (NamingException e) {
-                    _logger.log(Level.WARNING, "ldaprealm.exception", e);
+                    _logger.log(WARNING, "ldaprealm.exception", e);
                 }
             }
         }
+
         return null;
     }
 
@@ -385,32 +401,36 @@ public final class LDAPRealm extends IASRealm {
      */
     @Override
     public Enumeration<String> getGroupNames(String username) throws InvalidOperationException, NoSuchUserException {
-        Vector<String> v = groupCache.get(username);
-        if (v == null) {
-            //Note : assuming the username is a userDN here
+        Vector<String> groupNames = groupCache.get(username);
+        if (groupNames == null) {
+            // Note : assuming the username is a userDN here
             List<String> searchedGrps = getGroups(username);
             if (searchedGrps != null) {
                 return Collections.enumeration(searchedGrps);
             }
-            if (_logger.isLoggable(Level.FINE)) {
-                _logger.log(Level.FINE, "No groups available for: " + username);
+
+            if (_logger.isLoggable(FINE)) {
+                _logger.log(FINE, "No groups available for: " + username);
             }
-            // we don't load group here as we need to bind ctx to user with
+
+            // We don't load group here as we need to bind ctx to user with
             // password before doing that and password is not available here
             return emptyVector.elements();
         }
+
         if (groupMapper != null) {
             Vector<String> ret = new Vector<>();
-            ret.addAll(v);
+            ret.addAll(groupNames);
             ArrayList<String> result = new ArrayList<>();
-            for (String grp : v) {
+            for (String grp : groupNames) {
                 ArrayList<String> tmp = this.getMappedGroupNames(grp);
                 result.addAll(tmp);
             }
             ret.addAll(result);
             return ret.elements();
         }
-        return v.elements();
+
+        return groupNames.elements();
     }
 
     /**
@@ -421,11 +441,12 @@ public final class LDAPRealm extends IASRealm {
      *
      */
     private void setGroupNames(String username, String[] groups) {
-        Vector<String> v = new Vector<>(groups.length);
+        Vector<String> groupNames = new Vector<>(groups.length);
         for (String group : groups) {
-            v.add(group);
+            groupNames.add(group);
         }
-        groupCache.put(username, v);
+
+        groupCache.put(username, groupNames);
     }
 
     /**
@@ -433,16 +454,17 @@ public final class LDAPRealm extends IASRealm {
      *
      */
     public String[] findAndBind(String _username, char[] _password) throws LoginException {
-        // do search for user, substituting %s for username
+        // Do search for user, substituting %s for username
         _username = RFC2254Encode(_username);
-        StringBuffer sb = new StringBuffer(getProperty(PARAM_SEARCH_FILTER));
-        substitute(sb, SUBST_SUBJECT_NAME, _username);
-        String userid = sb.toString();
 
-        // attempt to bind as the user
+        StringBuilder userIdBuilder = new StringBuilder(getProperty(PARAM_SEARCH_FILTER));
+        substitute(userIdBuilder, SUBST_SUBJECT_NAME, _username);
+        String userid = userIdBuilder.toString();
+
+        // Attempt to bind as the user
         DirContext ctx = null;
         String srcFilter = null;
-        String[] grpList = null;
+        String[] groups = null;
 
         String dynFilter = null;
         String dynMember = getProperty(PARAM_DYNAMIC_GRP_TARGET);
@@ -450,8 +472,7 @@ public final class LDAPRealm extends IASRealm {
             ctx = new InitialDirContext(getLdapBindProps());
             String realUserDN = userSearch(ctx, getProperty(PARAM_USERDN), userid);
             if (realUserDN == null) {
-                String msg = sm.getString("ldaprealm.usernotfound", _username);
-                throw new LoginException(msg);
+                throw new LoginException(sm.getString("ldaprealm.usernotfound", _username));
             }
 
             boolean bindSuccessful = bindAsUser(realUserDN, _password);
@@ -461,23 +482,23 @@ public final class LDAPRealm extends IASRealm {
             }
 
             // search groups using above connection, substituting %d (and %s)
-            sb = new StringBuffer(getProperty(PARAM_GRP_SEARCH_FILTER));
-            StringBuffer dynSb = new StringBuffer(getProperty(PARAM_DYNAMIC_GRP_FILTER));
+            StringBuilder srcFilterBuilder = new StringBuilder(getProperty(PARAM_GRP_SEARCH_FILTER));
+            StringBuilder dynamicFilterBuilder = new StringBuilder(getProperty(PARAM_DYNAMIC_GRP_FILTER));
 
-            substitute(sb, SUBST_SUBJECT_NAME, _username);
-            substitute(sb, SUBST_SUBJECT_DN, realUserDN);
-            substitute(dynSb, SUBST_SUBJECT_NAME, _username);
-            substitute(dynSb, SUBST_SUBJECT_DN, realUserDN);
+            substitute(srcFilterBuilder, SUBST_SUBJECT_NAME, _username);
+            substitute(srcFilterBuilder, SUBST_SUBJECT_DN, realUserDN);
+            substitute(dynamicFilterBuilder, SUBST_SUBJECT_NAME, _username);
+            substitute(dynamicFilterBuilder, SUBST_SUBJECT_DN, realUserDN);
 
-            srcFilter = sb.toString();
-            dynFilter = dynSb.toString();
+            srcFilter = srcFilterBuilder.toString();
+            dynFilter = dynamicFilterBuilder.toString();
             ArrayList<String> groupsList = new ArrayList<>();
             groupsList.addAll(groupSearch(ctx, getProperty(PARAM_GRPDN), srcFilter, getProperty(PARAM_GRP_TARGET)));
-            // search filter is constructed internally as
-            // as a groupofURLS
+
+            // Search filter is constructed internally as as a groupofURLS
             groupsList.addAll(dynamicGroupSearch(ctx, getProperty(PARAM_GRPDN), dynMember, dynFilter, getProperty(PARAM_GRP_TARGET)));
-            grpList = new String[groupsList.size()];
-            groupsList.toArray(grpList);
+            groups = new String[groupsList.size()];
+            groupsList.toArray(groups);
         } catch (Exception e) {
             LoginException le = new LoginException(e.toString());
             le.initCause(e);
@@ -492,30 +513,33 @@ public final class LDAPRealm extends IASRealm {
             }
         }
 
-        if (_logger.isLoggable(Level.FINE)) {
-            _logger.log(Level.FINE, "LDAP:Group search filter: " + srcFilter);
-            StringBuffer gb = new StringBuffer();
+        if (_logger.isLoggable(FINE)) {
+            _logger.log(FINE, "LDAP:Group search filter: " + srcFilter);
+
+            StringBuilder gb = new StringBuilder();
             gb.append("Group memberships found: ");
-            if (grpList.length > 0) {
-                for (String element : grpList) {
+            if (groups.length > 0) {
+                for (String element : groups) {
                     gb.append(" " + element);
                 }
             } else {
                 gb.append("(null)");
             }
-            if (_logger.isLoggable(Level.FINE)) {
-                _logger.log(Level.FINE, "LDAP: " + gb.toString());
+
+            if (_logger.isLoggable(FINE)) {
+                _logger.log(FINE, "LDAP: " + gb.toString());
             }
         }
-        grpList = addAssignGroups(grpList);
-        grpList = this.addMappedGroupNames(grpList);
-        setGroupNames(_username, grpList);
 
-        if (_logger.isLoggable(Level.FINE)) {
-            _logger.log(Level.FINE, "LDAP: login succeeded for: " + _username);
+        groups = addAssignGroups(groups);
+        groups = this.addMappedGroupNames(groups);
+        setGroupNames(_username, groups);
+
+        if (_logger.isLoggable(FINE)) {
+            _logger.log(FINE, "LDAP: login succeeded for: " + _username);
         }
 
-        return grpList;
+        return groups;
     }
 
     /**
@@ -523,24 +547,24 @@ public final class LDAPRealm extends IASRealm {
      *
      */
     private String userSearch(DirContext ctx, String baseDN, String filter) {
-        if (_logger.isLoggable(Level.FINEST)) {
-            _logger.log(Level.FINE, "search: baseDN: " + baseDN + "  filter: " + filter);
+        if (_logger.isLoggable(FINEST)) {
+            _logger.log(FINE, "search: baseDN: " + baseDN + "  filter: " + filter);
         }
 
-        SearchControls ctls = new SearchControls();
-        ctls.setReturningAttributes(_dnOnly);
-        ctls.setSearchScope(SearchControls.SUBTREE_SCOPE);
-        ctls.setCountLimit(1);
+        SearchControls searchControls = new SearchControls();
+        searchControls.setReturningAttributes(_dnOnly);
+        searchControls.setSearchScope(SUBTREE_SCOPE);
+        searchControls.setCountLimit(1);
 
         NamingEnumeration<SearchResult> namingEnum = null;
         String foundDN;
         try {
-            namingEnum = ctx.search(baseDN, filter, ctls);
+            namingEnum = ctx.search(baseDN, filter, searchControls);
             if (namingEnum.hasMore()) {
                 SearchResult res = namingEnum.next();
 
                 StringBuffer sb = new StringBuffer();
-                //for dn name with '/'
+                // For dn name with '/'
                 CompositeName compDN = new CompositeName(res.getName());
                 String ldapDN = compDN.get(0);
                 sb.append(ldapDN);
@@ -550,15 +574,16 @@ public final class LDAPRealm extends IASRealm {
                     sb.append(baseDN);
                 }
                 foundDN = sb.toString();
-                if (_logger.isLoggable(Level.FINEST)) {
-                    _logger.log(Level.FINE, "Found user DN: " + foundDN);
+                if (_logger.isLoggable(FINEST)) {
+                    _logger.log(FINE, "Found user DN: " + foundDN);
                 }
+
                 return foundDN;
             }
             return null;
         } catch (Exception e) {
-            _logger.log(Level.WARNING, "ldaprealm.searcherror", filter);
-            _logger.log(Level.WARNING, "security.exception", e);
+            _logger.log(WARNING, "ldaprealm.searcherror", filter);
+            _logger.log(WARNING, "security.exception", e);
             return null;
         } finally {
             if (namingEnum != null) {
@@ -577,17 +602,17 @@ public final class LDAPRealm extends IASRealm {
     private boolean bindAsUser(String bindDN, char[] password) {
         boolean bindSuccessful = false;
 
-        Properties p = getLdapBindProps();
+        Properties ldapProperties = getLdapBindProps();
 
-        p.put(Context.SECURITY_PRINCIPAL, bindDN);
-        p.put(Context.SECURITY_CREDENTIALS, new String(password));
+        ldapProperties.put(SECURITY_PRINCIPAL, bindDN);
+        ldapProperties.put(SECURITY_CREDENTIALS, new String(password));
 
         DirContext ctx = null;
         try {
-            ctx = new InitialDirContext(p);
+            ctx = new InitialDirContext(ldapProperties);
             bindSuccessful = true;
         } catch (Exception e) {
-            if (_logger.isLoggable(Level.FINEST)) {
+            if (_logger.isLoggable(FINEST)) {
                 _logger.finest("Error binding to directory as: " + bindDN);
                 _logger.finest("Exception from JNDI: " + e.toString());
             }
@@ -599,6 +624,7 @@ public final class LDAPRealm extends IASRealm {
                 }
             }
         }
+
         return bindSuccessful;
     }
 
@@ -606,8 +632,7 @@ public final class LDAPRealm extends IASRealm {
      * Search for group membership using the given connection.
      *
      */
-    private List<String> dynamicGroupSearch(DirContext ctx, String baseDN, String memberOfAttr, String filter, String target)
-        throws NamingException {
+    private List<String> dynamicGroupSearch(DirContext ctx, String baseDN, String memberOfAttr, String filter, String target) throws NamingException {
         List<String> groupList = new ArrayList<>();
 
         String[] targets = new String[] { memberOfAttr };
@@ -615,9 +640,9 @@ public final class LDAPRealm extends IASRealm {
         try {
             SearchControls ctls = new SearchControls();
             ctls.setReturningAttributes(targets);
-            ctls.setSearchScope(SearchControls.SUBTREE_SCOPE);
-            //Set this to false to avoid objects and hence exposing ldap object
-            //injection.
+            ctls.setSearchScope(SUBTREE_SCOPE);
+
+            // Set this to false to avoid objects and hence exposing ldap object injection.
             ctls.setReturningObjFlag(false);
 
             NamingEnumeration<SearchResult> e = ctx.search(baseDN, filter, ctls);
@@ -638,8 +663,8 @@ public final class LDAPRealm extends IASRealm {
                 }
             }
         } catch (Exception e) {
-            _logger.log(Level.WARNING, "ldaprealm.searcherror", filter);
-            _logger.log(Level.WARNING, "security.exception", e);
+            _logger.log(WARNING, "ldaprealm.searcherror", filter);
+            _logger.log(WARNING, "security.exception", e);
         }
         return groupList;
     }
@@ -649,18 +674,18 @@ public final class LDAPRealm extends IASRealm {
      *
      */
     private List<String> groupSearch(DirContext ctx, String baseDN, String filter, String target) {
-        List<String> groupList = new ArrayList<>();
+        List<String> groups = new ArrayList<>();
 
         try {
             String[] targets = new String[1];
             targets[0] = target;
 
-            SearchControls ctls = new SearchControls();
-            ctls.setReturningAttributes(targets);
-            ctls.setSearchScope(SearchControls.SUBTREE_SCOPE);
+            SearchControls searchControls = new SearchControls();
+            searchControls.setReturningAttributes(targets);
+            searchControls.setSearchScope(SUBTREE_SCOPE);
 
             NamingEnumeration<SearchResult> e = ctx.search(baseDN,
-                filter.replaceAll(Matcher.quoteReplacement("\\"), Matcher.quoteReplacement("\\\\")), ctls);
+                filter.replaceAll(Matcher.quoteReplacement("\\"), Matcher.quoteReplacement("\\\\")), searchControls);
 
             while (e.hasMore()) {
                 SearchResult res = e.next();
@@ -668,23 +693,23 @@ public final class LDAPRealm extends IASRealm {
                 int sz = grpAttr.size();
                 for (int i = 0; i < sz; i++) {
                     String s = (String) grpAttr.get(i);
-                    groupList.add(s);
+                    groups.add(s);
                 }
             }
 
         } catch (Exception e) {
-            _logger.log(Level.WARNING, "ldaprealm.searcherror", filter);
-            _logger.log(Level.WARNING, "security.exception", e);
+            _logger.log(WARNING, "ldaprealm.searcherror", filter);
+            _logger.log(WARNING, "security.exception", e);
         }
 
-        return groupList;
+        return groups;
     }
 
     /**
      * Do string substitution. target is replaced by value for all occurences.
      *
      */
-    private static void substitute(StringBuffer sb, String target, String value) {
+    private static void substitute(StringBuilder sb, String target, String value) {
         int i = sb.indexOf(target);
         while (i >= 0) {
             sb.replace(i, i + target.length(), value);
@@ -699,7 +724,6 @@ public final class LDAPRealm extends IASRealm {
      * @return
      */
     private String RFC2254Encode(String inName) {
-
         int len = inName.length();
         StringBuilder buf = new StringBuilder(len);
         for (int i = 0; i < len; i++) {

